File: internal/integrations/jira/actions.go
package jira

import (
	"errors"
	"fmt"
	"regexp"

	"github.com/roee-hersh/hallpass/internal/catalog"
)

// action is one Jira permission key. global ones are checked with
// globalPermissions on the resource "global"; the rest are project
// permissions checked on project:<KEY> or issue:<KEY-N>.
type action struct {
	name, desc string
	global     bool
}

var actionList = []action{
	{name: "BROWSE_PROJECTS", desc: "view the project and its issues"},
	{name: "CREATE_ISSUES", desc: "create issues in the project"},
	{name: "EDIT_ISSUES", desc: "edit issues"},
	{name: "DELETE_ISSUES", desc: "delete issues"},
	{name: "ASSIGN_ISSUES", desc: "assign issues to users"},
	{name: "ASSIGNABLE_USER", desc: "be assigned issues"},
	{name: "TRANSITION_ISSUES", desc: "transition issues through the workflow"},
	{name: "RESOLVE_ISSUES", desc: "resolve and reopen issues, set fix versions"},
	{name: "CLOSE_ISSUES", desc: "close issues"},
	{name: "MOVE_ISSUES", desc: "move issues between projects or issue types"},
	{name: "LINK_ISSUES", desc: "link issues"},
	{name: "ADD_COMMENTS", desc: "add comments"},
	{name: "EDIT_ALL_COMMENTS", desc: "edit any comment"},
	{name: "DELETE_ALL_COMMENTS", desc: "delete any comment"},
	{name: "CREATE_ATTACHMENTS", desc: "attach files"},
	{name: "WORK_ON_ISSUES", desc: "log work on issues"},
	{name: "MANAGE_WATCHERS", desc: "manage the watcher list"},
	{name: "VIEW_VOTERS_AND_WATCHERS", desc: "view voters and watchers"},
	{name: "SCHEDULE_ISSUES", desc: "schedule issues (due date, rank)"},
	{name: "SET_ISSUE_SECURITY", desc: "set the security level of issues"},
	{name: "MANAGE_SPRINTS_PERMISSION", desc: "manage sprints"},
	{name: "ADMINISTER_PROJECTS", desc: "administer the project"},
	{name: "ADMINISTER", desc: "administer Jira (global)", global: true},
	{name: "SYSTEM_ADMIN", desc: "administer Jira system settings (global)", global: true},
	{name: "USER_PICKER", desc: "browse users and groups (global)", global: true},
	{name: "CREATE_SHARED_OBJECTS", desc: "share filters and dashboards (global)", global: true},
	{name: "MANAGE_GROUP_FILTER_SUBSCRIPTIONS", desc: "manage group filter subscriptions (global)", global: true},
	{name: "BULK_CHANGE", desc: "make bulk changes (global)", global: true},
}

var actions = func() map[string]action {
	m := map[string]action{}
	for _, a := range actionList {
		m[a.name] = a
	}
	return m
}()

var (
	projectKeyRe = regexp.MustCompile(`^[A-Z][A-Z0-9_]{1,9}$`)
	issueKeyRe   = regexp.MustCompile(`^[A-Z][A-Z0-9_]{1,9}-[1-9][0-9]*$`)
)

type resourceKind int

const (
	resGlobal resourceKind = iota
	resProject
	resIssue
)

// resource is a parsed project:<KEY>, issue:<KEY-N> or global.
type resource struct {
	kind resourceKind
	key  string
}

func (r resource) describe() string {
	switch r.kind {
	case resProject:
		return "project " + r.key
	case resIssue:
		return "issue " + r.key
	}
	return "this site"
}

func parseResource(res catalog.Resource) (resource, error) {
	switch res.Type {
	case "global":
		if res.ID != "" {
			return resource{}, errors.New("global takes no id")
		}
		return resource{kind: resGlobal}, nil
	case "project":
		if !projectKeyRe.MatchString(res.ID) {
			return resource{}, fmt.Errorf("project key %q must match %s", res.ID, projectKeyRe)
		}
		return resource{kind: resProject, key: res.ID}, nil
	case "issue":
		if !issueKeyRe.MatchString(res.ID) {
			return resource{}, fmt.Errorf("issue key %q must match %s", res.ID, issueKeyRe)
		}
		return resource{kind: resIssue, key: res.ID}, nil
	default:
		return resource{}, fmt.Errorf("resource type %q; use project:<KEY>, issue:<KEY-N> or global", res.Type)
	}
}

File: hallpass-py/src/hallpass/integrations/jira/actions.py
"""The jira actions (native permission keys) and resource parsing."""

from __future__ import annotations

import re
from dataclasses import dataclass
from enum import IntEnum

from hallpass.core.catalog import Resource
from hallpass.core.errors import go_quote

__all__ = [
    "ACTIONS",
    "ACTION_LIST",
    "ISSUE_KEY_PATTERN",
    "ISSUE_KEY_RE",
    "PROJECT_KEY_PATTERN",
    "PROJECT_KEY_RE",
    "JiraAction",
    "JiraResource",
    "ResKind",
    "parse_resource",
]


@dataclass(frozen=True)
class JiraAction:
    """One Jira permission key. global ones are checked with
    globalPermissions on the resource "global"; the rest are project
    permissions checked on project:<KEY> or issue:<KEY-N>."""

    name: str
    desc: str
    global_: bool = False


ACTION_LIST: tuple[JiraAction, ...] = (
    JiraAction("BROWSE_PROJECTS", "view the project and its issues"),
    JiraAction("CREATE_ISSUES", "create issues in the project"),
    JiraAction("EDIT_ISSUES", "edit issues"),
    JiraAction("DELETE_ISSUES", "delete issues"),
    JiraAction("ASSIGN_ISSUES", "assign issues to users"),
    JiraAction("ASSIGNABLE_USER", "be assigned issues"),
    JiraAction("TRANSITION_ISSUES", "transition issues through the workflow"),
    JiraAction("RESOLVE_ISSUES", "resolve and reopen issues, set fix versions"),
    JiraAction("CLOSE_ISSUES", "close issues"),
    JiraAction("MOVE_ISSUES", "move issues between projects or issue types"),
    JiraAction("LINK_ISSUES", "link issues"),
    JiraAction("ADD_COMMENTS", "add comments"),
    JiraAction("EDIT_ALL_COMMENTS", "edit any comment"),
    JiraAction("DELETE_ALL_COMMENTS", "delete any comment"),
    JiraAction("CREATE_ATTACHMENTS", "attach files"),
    JiraAction("WORK_ON_ISSUES", "log work on issues"),
    JiraAction("MANAGE_WATCHERS", "manage the watcher list"),
    JiraAction("VIEW_VOTERS_AND_WATCHERS", "view voters and watchers"),
    JiraAction("SCHEDULE_ISSUES", "schedule issues (due date, rank)"),
    JiraAction("SET_ISSUE_SECURITY", "set the security level of issues"),
    JiraAction("MANAGE_SPRINTS_PERMISSION", "manage sprints"),
    JiraAction("ADMINISTER_PROJECTS", "administer the project"),
    JiraAction("ADMINISTER", "administer Jira (global)", global_=True),
    JiraAction("SYSTEM_ADMIN", "administer Jira system settings (global)", global_=True),
    JiraAction("USER_PICKER", "browse users and groups (global)", global_=True),
    JiraAction("CREATE_SHARED_OBJECTS", "share filters and dashboards (global)", global_=True),
    JiraAction("MANAGE_GROUP_FILTER_SUBSCRIPTIONS", "manage group filter subscriptions (global)", global_=True),
    JiraAction("BULK_CHANGE", "make bulk changes (global)", global_=True),
)

ACTIONS: dict[str, JiraAction] = {a.name: a for a in ACTION_LIST}

# The Go patterns, as error messages print them. The compiled forms are
# used with fullmatch: Go's $ matches only at the end of the text, while
# Python's also matches before a trailing newline.
PROJECT_KEY_PATTERN = "^[A-Z][A-Z0-9_]{1,9}$"
ISSUE_KEY_PATTERN = "^[A-Z][A-Z0-9_]{1,9}-[1-9][0-9]*$"
PROJECT_KEY_RE = re.compile(r"[A-Z][A-Z0-9_]{1,9}")
ISSUE_KEY_RE = re.compile(r"[A-Z][A-Z0-9_]{1,9}-[1-9][0-9]*")


class ResKind(IntEnum):
    GLOBAL = 0
    PROJECT = 1
    ISSUE = 2


@dataclass(frozen=True)
class JiraResource:
    """A parsed project:<KEY>, issue:<KEY-N> or global."""

    kind: ResKind
    key: str = ""

    def describe(self) -> str:
        if self.kind == ResKind.PROJECT:
            return "project " + self.key
        if self.kind == ResKind.ISSUE:
            return "issue " + self.key
        return "this site"


def parse_resource(res: Resource) -> JiraResource:
    """The resource as a Jira target; ValueError with Go's message otherwise."""
    if res.type == "global":
        if res.id != "":
            raise ValueError("global takes no id")
        return JiraResource(ResKind.GLOBAL)
    if res.type == "project":
        if not PROJECT_KEY_RE.fullmatch(res.id):
            raise ValueError(f"project key {go_quote(res.id)} must match {PROJECT_KEY_PATTERN}")
        return JiraResource(ResKind.PROJECT, res.id)
    if res.type == "issue":
        if not ISSUE_KEY_RE.fullmatch(res.id):
            raise ValueError(f"issue key {go_quote(res.id)} must match {ISSUE_KEY_PATTERN}")
        return JiraResource(ResKind.ISSUE, res.id)
    raise ValueError(f"resource type {go_quote(res.type)}; use project:<KEY>, issue:<KEY-N> or global")
